### ISSUE TITLE
Completion stops working in eshell when typing quote characters

### ISSUE THREAD

**Issue by snyball** [doomemacs/doomemacs#3817]
**What did you expect to happen?**
For `company-mode` auto-completion to keep working.

**What actually happened?**
The `company-mode` completion window stops popping up.

**Additional details:**

This is the error-message written to the message-buffer when completions stop working:

```
Completion function pcomplete-completions-at-point uses a deprecated calling convention
```

Running with `debug-on-error t` does not produce any additional backtrace.

**Steps to reproduce:**
1. `SPC o E`
2. Confirm that `company-mode` completions are working
3. Type `<anything> '`

After typing the `'` character you should see a `TAB` inserted, the echo are should display the error message about `deprecated calling convention`. Completions should now no longer be working correctly.

I attempted to remove the `pcomplete-completions-at-point` function from `completion-at-point-functions`, but this resulted in no completions. Does the `fish-completion` package use `pcomplete`?

```emacs-lisp
(add-hook! (eshell-mode) :append
  (remove-hook 'completion-at-point-functions #'pcomplete-completions-at-point t))
```

I also tried `(setq company-backends '(company-pcomplete))` (from `modules/term/eshell/autoload/company.el`.) But that backend throws the following error when typing either `"` or `'` (also inserts a `TAB`.)

<details>

```
Debugger entered--Lisp error: (no-catch pcompleted t)
  throw(pcompleted t)
  eshell-complete-parse-arguments()
  pcomplete-parse-arguments(nil)
  pcomplete-completions()
  (let* ((pcomplete-stub) pcomplete-seen pcomplete-norm-func pcomplete-args pcomplete-last pcomplete-index (pcomplete-autolist pcomplete-autolist) (pcomplete-suffix-list pcomplete-suffix-list)) (pcomplete-completions) (buffer-substring (pcomplete-begin) (point)))
  (with-no-warnings (let* ((pcomplete-stub) pcomplete-seen pcomplete-norm-func pcomplete-args pcomplete-last pcomplete-index (pcomplete-autolist pcomplete-autolist) (pcomplete-suffix-list pcomplete-suffix-list)) (pcomplete-completions) (buffer-substring (pcomplete-begin) (point))))
  company-pcomplete--prefix()
  (progn (company-pcomplete--prefix))
  (if (company-pcomplete-available) (progn (company-pcomplete--prefix)))
  (cond ((eql command 'interactive) (company-begin-backend 'company-pcomplete)) ((eql command 'prefix) (if (company-pcomplete-available) (progn (company-pcomplete--prefix)))) ((eql command 'candidates) (company-pcomplete--candidates)) ((eql command 'sorted) t))
  company-pcomplete(prefix)
  apply(company-pcomplete prefix)
  company-call-backend-raw(prefix)
  company--force-sync(company-call-backend-raw (prefix) company-pcomplete)
  company-call-backend(prefix)
  company--begin-new()
  company--perform()
  company-auto-begin()
  company-idle-begin(#<buffer *doom:eshell*> #<window 3 on *doom:eshell*> 268 6237)
  apply(company-idle-begin (#<buffer *doom:eshell*> #<window 3 on *doom:eshell*> 268 6237))
  timer-event-handler([t 24387 9335 232068 nil company-idle-begin (#<buffer *doom:eshell*> #<window 3 on *doom:eshell*> 268 6237) nil 337000])
```
</details>

**System information:**
<details>

```
emacs   version    28.0.50
        features   JPEG TIFF GIF PNG RSVG CAIRO SOUND GPM DBUS GLIB NOTIFY INOTIFY ACL GNUTLS LIBXML2 FREETYPE HARFBUZZ LIBOTF ZLIB TOOLKIT_SCROLL_BARS GTK3 PGTK XIM MODULES NATIVE_COMP THREADS XWIDGETS LIBSYSTEMD JSON PDUMPER LCMS2
        build      aug. 18, 2020
        buildopts  (--with-nativecomp --with-gnutls --without-gconf --without-gsettings --with-rsvg --with-xwidgets --with-webkit --with-pgtk --with-cairo --without-xaw3d --with-mailutils --prefix=/home/sanoj/.local/opt 'CFLAGS=-O3 -mtune=native -march=native -fomit-frame-pointer')
        windowsys  pgtk
        daemonp    daemon
doom    version    2.0.9
        build      HEAD -> develop, origin/develop, origin/HEAD e54bb170c 2020-08-21 01:26:42 -0400
        dir        ~/.config/doom/
system  type       gnu/linux
        config     x86_64-pc-linux-gnu
        shell      /usr/bin/fish
        uname      Linux 5.4.52-1-MANJARO #1 SMP PREEMPT Thu Jul 16 16:07:11 UTC 2020 x86_64
        path       (~/.local/bin ~/.local/share/gem/bin ~/.local/share/cargo/bin ~/.local/go/bin /usr/local/sbin /usr/local/bin /usr/bin ~/.local/opt/bin/ /var/lib/flatpak/exports/bin /usr/lib/jvm/default/bin /usr/bin/site_perl /usr/bin/vendor_perl /usr/bin/core_perl ~/.local/opt/libexec/emacs/28.0.50/x86_64-pc-linux-gnu)
config  envfile    nil
        elc-files  0
        modules    (:completion (company +childframe) (ivy +childframe +prescient +icons) :ui doom doom-dashboard doom-quit hl-todo modeline popup treemacs unicode vc-gutter window-select workspaces hydra zen ophints :editor (evil +everywhere) file-templates fold lispy multiple-cursors rotate-text snippets :emacs undo (dired +ranger +icons) electric vc :term eshell shell term vterm :checkers (syntax +childframe) (spell +childframe) :tools ein (eval +overlay) gist (lookup +docsets) lsp (magit +forge) make pdf rgb docker :lang (rust +lsp) (cc +lsp) (ruby +lsp) (haskell +intero +lsp) (python +lsp) (java +lsp) (javascript +lsp) lua emacs-lisp scheme ess julia (sh +fish) data latex markdown (org +dragndrop +ipython +pandoc +gnuplot +pretty +roam +present) web yaml json :app calendar :email mu4e :config (default +smartparens))
        packages   ((nginx-mode) (vimrc-mode) (smart-semicolon) (ox-json) (strace-mode) (meson-mode) (pkgbuild-mode) (ssh-config-mode) (string-inflection) (ace-link) (jq-mode) (systemd) (pcre2el) (matrix-client) (hydra-posframe :recipe (:repo https://github.com/Ladicle/hydra-posframe.git)))
        unpin      (n/a)
        elpa       (n/a)
```
</details>

**Comment by snyball** [comment]
It seems like this is all resting on a good ol' pile of horrifying hacks: https://github.com/company-mode/company-mode/issues/409#issuecomment-434820576

@hlissner Should doom include workarounds for this or just wait for upstream (note: the linked bug report is from 2018.)

**Comment by snyball** [comment]
From the `pcomplete` source:

```emacs-lisp
  ;; FIXME: it only completes the text before point, whereas the
  ;; standard UI may also consider text after point.
  ;; FIXME: the `pcomplete' UI may be used internally during
  ;; pcomplete-completions and then throw to `pcompleted', thus
  ;; imposing the pcomplete UI over the standard UI.
```

The following hack at least makes the completion work again after executing a command:

```emacs-lisp
(defun eshell-completion-hotfix ()
  (remove-hook 'completion--capf-misbehave-funs
               #'pcomplete-completions-at-point))
(add-hook 'eshell-post-command-hook #'eshell-completion-hotfix)
```

The issue seems to be that `eshell-complete-parse-arguments` expects to signal completion with `throw`, which isn't supported by either `company-mode` or `capf`. The following patch to `eshell-complete-parse-arguments` fixes the `company-mode` problems.

<details>

```emacs-lisp
(defun eshell-complete-parse-arguments ()
  "Parse the command line arguments for `pcomplete-argument'."
  (catch 'pcompleted
    (unless (and eshell-no-completion-during-jobs
                 (eshell-interactive-process))
      (let ((end (point-marker))
            (begin (save-excursion (eshell-bol) (point)))
            (posns (list t))
            args delim)
        (when (memq this-command '(pcomplete-expand
                                   pcomplete-expand-and-complete))
          (run-hook-with-args 'eshell-expand-input-functions begin end)
          (if (= begin end)
              (end-of-line))
          (setq end (point-marker)))
        (if (setq delim
                  (catch 'eshell-incomplete
                    (ignore
                     (setq args (eshell-parse-arguments begin end)))))
            (cond ((memq (car delim) '(?\{ ?\<))
                   (setq begin (1+ (cadr delim))
                         args (eshell-parse-arguments begin end)))
                  (t (throw 'pcompleted `(nil ,begin ,end)))))
        (when (get-text-property (1- end) 'comment)
          (throw 'pcompleted `(nil ,begin ,end)))
        (let ((pos begin))
          (while (< pos end)
            (if (get-text-property pos 'arg-begin)
                (nconc posns (list pos)))
            (setq pos (1+ pos))))
        (setq posns (cdr posns))
        (cl-assert (= (length args) (length posns)))
        (let ((a args)
              (i 0)
              l)
          (while a
            (if (and (consp (car a))
                     (eq (caar a) 'eshell-operator))
                (setq l i))
            (setq a (cdr a) i (1+ i)))
          (and l
               (setq args (nthcdr (1+ l) args)
                     posns (nthcdr (1+ l) posns))))
        (cl-assert (= (length args) (length posns)))
        (when (and args (eq (char-syntax (char-before end)) ? )
                   (not (eq (char-before (1- end)) ?\\)))
          (nconc args (list ""))
          (nconc posns (list (point))))
        (cons (mapcar
               (function
                (lambda (arg)
                  (let ((val
                         (if (listp arg)
                             (let ((result
                                    (eshell-do-eval
                                     (list 'eshell-commands arg) t)))
                               (cl-assert (eq (car result) 'quote))
                               (cadr result))
                           arg)))
                    (if (numberp val)
                        (setq val (number-to-string val)))
                    (or val ""))))
               args)
              posns)))))
```
</details>

Then, I just set:

```emacs-lisp
;; The `eshell-complete-parse-arguments' function has been patched to fail on lisp expressions, so that we can just
;; fall back on the `elisp-completion-at-point` backend.
(setq completion-at-point-functions '(pcomplete-completions-at-point elisp-completion-at-point))
```

Considering how old the bug seems to be, and the lack of response from upstream to [this](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=40651) bug report, I think it might be worthwhile to add these patches to Doom, activated only if the `company` module is enabled. @hlissner  If you agree that it should be patched in Doom I can submit a pull-request in the morning.

**Comment by snyball** [comment]
Pinging @hlissner 

The following patch has worked for me:

<details>

```emacs-lisp
;;; eshell-company-patch.el -*- lexical-binding: t; -*-

;; Copyright (C) 1999-2020 Free Software Foundation, Inc.

;; Author: John Wiegley <johnw@gnu.org>

;; This file is NOT part of GNU Emacs.

;; GNU Emacs is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; GNU Emacs is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:
;; The `eshell-complete-parse-arguments' function has been modified to work
;; better with the `company-mode' package.
;;
;; The function was originally taken from the source file `em-cmpl',
;; for which original source can be found here:
;;   https://git.savannah.gnu.org/cgit/emacs.git/tree/lisp/eshell/em-cmpl.el
;;
;; Notable changes:
;;   - Catching of pcompleted inside `eshell-complete-parse-arguments'
;;   - Removal of some `throw's
;;   - Changing the remaining `throw's second parameter to be:
;;       `(nil ,begin ,end)
;;     rather than `t'.

(defun doom-eshell-overrides ()
  (defun eshell-complete-parse-arguments ()
    "Parse the command line arguments for `pcomplete-argument'."
    (catch 'pcompleted
      (unless (and eshell-no-completion-during-jobs
                   (eshell-interactive-process))
        (let ((end (point-marker))
              (begin (save-excursion (eshell-bol) (point)))
              (posns (list t))
              args delim)
          (when (memq this-command '(pcomplete-expand
                                     pcomplete-expand-and-complete))
            (run-hook-with-args 'eshell-expand-input-functions begin end)
            (if (= begin end)
                (end-of-line))
            (setq end (point-marker)))
          (if (setq delim
                    (catch 'eshell-incomplete
                      (ignore
                       (setq args (eshell-parse-arguments begin end)))))
              (cond ((memq (car delim) '(?\{ ?\<))
                     (setq begin (1+ (cadr delim))
                           args (eshell-parse-arguments begin end)))
                    (t (throw 'pcompleted `(nil ,begin ,end)))))
          (when (get-text-property (1- end) 'comment)
            (throw 'pcompleted `(nil ,begin ,end)))
          (let ((pos begin))
            (while (< pos end)
              (if (get-text-property pos 'arg-begin)
                  (nconc posns (list pos)))
              (setq pos (1+ pos))))
          (setq posns (cdr posns))
          (cl-assert (= (length args) (length posns)))
          (let ((a args)
                (i 0)
                l)
            (while a
              (if (and (consp (car a))
                       (eq (caar a) 'eshell-operator))
                  (setq l i))
              (setq a (cdr a) i (1+ i)))
            (and l
                 (setq args (nthcdr (1+ l) args)
                       posns (nthcdr (1+ l) posns))))
          (cl-assert (= (length args) (length posns)))
          (when (and args (eq (char-syntax (char-before end)) ? )
                     (not (eq (char-before (1- end)) ?\\)))
            (nconc args (list ""))
            (nconc posns (list (point))))
          (cons (mapcar
                 (function
                  (lambda (arg)
                    (let ((val
                           (if (listp arg)
                               (let ((result
                                      (eshell-do-eval
                                       (list 'eshell-commands arg) t)))
                                 (cl-assert (eq (car result) 'quote))
                                 (cadr result))
                             arg)))
                      (if (numberp val)
                          (setq val (number-to-string val)))
                      (or val ""))))
                 args)
                posns))))))

(when (and (featurep! :completion company)
           (featurep! :term eshell))
  (add-hook 'eshell-cmpl-load-hook #'doom-eshell-overrides))

(provide 'eshell-company-patch)
;; eshell-company-patch.el ends here
```
</details>

It's only activated if both `company` and `eshell` are enabled, as `eshell` is _really_ broken with `company` otherwise. It complies with the GPL by listing changes and linking to the original source, as well as preserving the original license and FSF copyright notice.

**Comment by hlissner** [comment]
> Considering how old the bug seems to be, and the lack of response from upstream to this bug report, I think it might be worthwhile to add these patches to Doom, activated only if the company module is enabled.

Seems we have no choice.

----

I cannot find where `eshell-complete-parse-arguments` is defined on 27.1. I'd rather reduce your patch to an advice, but need to see how exactly your patched version is different from the original to do so.

**Comment by snyball** [comment]
@hlissner Here's a diff:

<details>

```diff
diff -u /tmp/em-cmpl-eshell-complete-parse-arguments.orig.el /tmp/my-eshell-complete-parse-arguments.new.el
--- /tmp/em-cmpl-eshell-complete-parse-arguments.orig.el	2020-08-28 13:05:18.185416575 +0200
+++ /tmp/my-eshell-complete-parse-arguments.new.el	2020-08-28 13:05:57.378570724 +0200
@@ -1,71 +1,64 @@
 (defun eshell-complete-parse-arguments ()
   "Parse the command line arguments for `pcomplete-argument'."
-  (when (and eshell-no-completion-during-jobs
-             (eshell-interactive-process))
-    (insert-and-inherit "\t")
-    (throw 'pcompleted t))
-  (let ((end (point-marker))
-        (begin (save-excursion (eshell-bol) (point)))
-        (posns (list t))
-        args delim)
-    (when (memq this-command '(pcomplete-expand
-                               pcomplete-expand-and-complete))
-      (run-hook-with-args 'eshell-expand-input-functions begin end)
-      (if (= begin end)
-          (end-of-line))
-      (setq end (point-marker)))
-    (if (setq delim
-              (catch 'eshell-incomplete
-                (ignore
-                 (setq args (eshell-parse-arguments begin end)))))
-        (cond ((memq (car delim) '(?\{ ?\<))
-               (setq begin (1+ (cadr delim))
-                     args (eshell-parse-arguments begin end)))
-              ((eq (car delim) ?\()
-               (eshell-complete-lisp-symbol)
-               (throw 'pcompleted t))
-              (t
-               (insert-and-inherit "\t")
-               (throw 'pcompleted t))))
-    (when (get-text-property (1- end) 'comment)
-      (insert-and-inherit "\t")
-      (throw 'pcompleted t))
-    (let ((pos begin))
-      (while (< pos end)
-        (if (get-text-property pos 'arg-begin)
-            (nconc posns (list pos)))
-        (setq pos (1+ pos))))
-    (setq posns (cdr posns))
-    (cl-assert (= (length args) (length posns)))
-    (let ((a args)
-          (i 0)
-          l)
-      (while a
-        (if (and (consp (car a))
-                 (eq (caar a) 'eshell-operator))
-            (setq l i))
-        (setq a (cdr a) i (1+ i)))
-      (and l
-           (setq args (nthcdr (1+ l) args)
-                 posns (nthcdr (1+ l) posns))))
-    (cl-assert (= (length args) (length posns)))
-    (when (and args (eq (char-syntax (char-before end)) ? )
-               (not (eq (char-before (1- end)) ?\\)))
-      (nconc args (list ""))
-      (nconc posns (list (point))))
-    (cons (mapcar
-           (function
-            (lambda (arg)
-              (let ((val
-                     (if (listp arg)
-                         (let ((result
-                                (eshell-do-eval
-                                 (list 'eshell-commands arg) t)))
-                           (cl-assert (eq (car result) 'quote))
-                           (cadr result))
-                       arg)))
-                (if (numberp val)
-                    (setq val (number-to-string val)))
-                (or val ""))))
-           args)
-          posns)))
+  (catch 'pcompleted
+    (unless (and eshell-no-completion-during-jobs
+                 (eshell-interactive-process))
+      (let ((end (point-marker))
+            (begin (save-excursion (eshell-bol) (point)))
+            (posns (list t))
+            args delim)
+        (when (memq this-command '(pcomplete-expand
+                                   pcomplete-expand-and-complete))
+          (run-hook-with-args 'eshell-expand-input-functions begin end)
+          (if (= begin end)
+              (end-of-line))
+          (setq end (point-marker)))
+        (if (setq delim
+                  (catch 'eshell-incomplete
+                    (ignore
+                     (setq args (eshell-parse-arguments begin end)))))
+            (cond ((memq (car delim) '(?\{ ?\<))
+                   (setq begin (1+ (cadr delim))
+                         args (eshell-parse-arguments begin end)))
+                  (t (throw 'pcompleted `(nil ,begin ,end)))))
+        (when (get-text-property (1- end) 'comment)
+          (throw 'pcompleted `(nil ,begin ,end)))
+        (let ((pos begin))
+          (while (< pos end)
+            (if (get-text-property pos 'arg-begin)
+                (nconc posns (list pos)))
+            (setq pos (1+ pos))))
+        (setq posns (cdr posns))
+        (cl-assert (= (length args) (length posns)))
+        (let ((a args)
+              (i 0)
+              l)
+          (while a
+            (if (and (consp (car a))
+                     (eq (caar a) 'eshell-operator))
+                (setq l i))
+            (setq a (cdr a) i (1+ i)))
+          (and l
+               (setq args (nthcdr (1+ l) args)
+                     posns (nthcdr (1+ l) posns))))
+        (cl-assert (= (length args) (length posns)))
+        (when (and args (eq (char-syntax (char-before end)) ? )
+                   (not (eq (char-before (1- end)) ?\\)))
+          (nconc args (list ""))
+          (nconc posns (list (point))))
+        (cons (mapcar
+               (function
+                (lambda (arg)
+                  (let ((val
+                         (if (listp arg)
+                             (let ((result
+                                    (eshell-do-eval
+                                     (list 'eshell-commands arg) t)))
+                               (cl-assert (eq (car result) 'quote))
+                               (cadr result))
+                           arg)))
+                    (if (numberp val)
+                        (setq val (number-to-string val)))
+                    (or val ""))))
+               args)
+              posns)))))
```
</details>

You might be able to reduce it to an advice, but I'm not sure about the `(insert-and-inherit "\t")` stuff. It was (presumably) added with the assumption that people would use `TAB` to initiate auto-completion, but becomes a nuisance when using automatic timer-based completion popups (or another key-binding.)

**Comment by minad** [comment]
I just investigated this issue in the context of my Corfu completion package which exclusively builds upon Capfs (`completion-at-point-functions`). While the modern Capf API is okay, there seems to be some legacy involved. Some Capfs registered in the `completion-at-point-functions` list misbehave. Instead of returning a completion table which can then be used by the frontend, e.g., Company, these Capfs perform completion right away.

A particular problem is the function `pcomplete-completions-at-point` which sometimes behaves well and returns a completion table and sometimes just completes (inserts a string as in this case). As a consequence Company complains about the deprecated behavior. Since the Capf behaves well most of the time, I don't like sledgehammer solution of simply deactivating the Capf.

Of course in the long term the issue should be fixed upstream and all these misbehaving Capfs should be replaced. In the meantime I am using Capf wrappers/transformers, which can turn bad Capfs into good Capfs by undoing their misbehavior. I collect a set of Capf wrappers in my [Cape](https://github.com/minad/cape) package. Not all of the wrappers have the purpose of "fixing" Capfs, some also augment the functionality of existing Capfs.

Regarding this issue see in particular `cape-repair-misbehaving-capf` (https://github.com/minad/cape#repair-misbehaving-capfs), which could help. It works well for me with Corfu and it also seems to work with Company but I haven't tested it extensively. I am not sure if this issue has been fixed in Doom (with the patch from https://github.com/hlissner/doom-emacs/issues/3817#issuecomment-681922391), but if not you could try `cape-repair-misbehaving-capf`. You don't have to depend on Cape. Instead just copy the function to the config directly, since it is a small and localized fix.

**Comment by minad** [comment]
There is some movement now, we may also see a fix in Company and upstream. Probably the fix won't be a proper fix for pcomplete which would be the ideal solution. But it makes sense to limit misbehaving capfs in the amount of damage they can do by setting the buffer to read-only and catching the error. In my Corfu UI I adoped such a workaround and Company may do the same (https://github.com/minad/corfu/blob/160bd53db56f279d87377a8068aab754ec4b5d0c/corfu.el#L1153-L1183).

See also the upstream discussion https://debbugs.gnu.org/cgi/bugreport.cgi?bug=50470

**Comment by danilevy1212** [comment]
Cou

> From the `pcomplete` source:
> 
> ```emacs-lisp
>   ;; FIXME: it only completes the text before point, whereas the
>   ;; standard UI may also consider text after point.
>   ;; FIXME: the `pcomplete' UI may be used internally during
>   ;; pcomplete-completions and then throw to `pcompleted', thus
>   ;; imposing the pcomplete UI over the standard UI.
> ```
> 
> The following hack at least makes the completion work again after executing a command:
> 
> ```emacs-lisp
> (defun eshell-completion-hotfix ()
>   (remove-hook 'completion--capf-misbehave-funs
>                #'pcomplete-completions-at-point))
> (add-hook 'eshell-post-command-hook #'eshell-completion-hotfix)
> ```
> 
> The issue seems to be that `eshell-complete-parse-arguments` expects to signal completion with `throw`, which isn't supported by either `company-mode` or `capf`. The following patch to `eshell-complete-parse-arguments` fixes the `company-mode` problems.
> 
> Then, I just set:
> 
> ```emacs-lisp
> ;; The `eshell-complete-parse-arguments' function has been patched to fail on lisp expressions, so that we can just
> ;; fall back on the `elisp-completion-at-point` backend.
> (setq completion-at-point-functions '(pcomplete-completions-at-point elisp-completion-at-point))
> ```
> 
> Considering how old the bug seems to be, and the lack of response from upstream to [this](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=40651) bug report, I think it might be worthwhile to add these patches to Doom, activated only if the `company` module is enabled. I can submit a pull-request if it's something you want to patch in Doom.

Could we make a PR of this? I am using it and it seems to work.

**Comment by minad** [comment]
@danilevy1212 The eshell hotfix is not a great idea. I recommend to use the advices provided by my cape package. See https://github.com/minad/corfu#completing-in-the-eshell-or-shell. These advices fix the issues properly. I believe that on Emacs 29 the issue is fixed at the root, but I haven't tested this yet.

**Comment by danilevy1212** [comment]
Thank you @minad, I'll definitely check it out

**Comment by danilevy1212** [comment]
@minad To use cape we would have to bring it down as a dependency which might be a bit excessive just for this fix, although if there is no problem with it I am for it as it is the easiest solution.

Otherwise I would have to rewrite the functions for our use. I do have a a question about one of them though:
```lisp
;;;###autoload
(defun cape-wrap-silent (capf)
  "Call CAPF and silence it (no messages, no errors)."
  (pcase (cape--silent (funcall capf))
    (`(,beg ,end ,table . ,plist)
     `(,beg ,end ,(cape--silent-table table) ,@plist))))
```

Not sure what `plist` is supposed to be in the definition? Is it something particular from `pcase` or from `cape`?

**Comment by minad** [comment]
@danilevy1212 The cape dependency is pretty small, but you could also copy the definitions. plist comes from the pcase pattern matching.

**Comment by danilevy1212** [comment]
Unfortunately, I can't seem to make time for this before the modules become their own repository, I have the fix working on my personal config, is just a matter of patching it into the eshell module:

```org
#+begin_src emacs-lisp :tangle packages.el
(package! cape)
#+end_src

Until [[https://github.com/doomemacs/doomemacs/issues/3817][completion stops working in eshell when typing quote characters]] is fixed, we apply the fix ourselves.  See [[https://github.com/doomemacs/doomemacs/issues/3817][here]].

#+begin_src emacs-lisp :tangle yes
(use-package! cape
  :init
  (after! eshell
    (advice-add 'pcomplete-completions-at-point :around #'cape-wrap-silent)
    (advice-add 'pcomplete-completions-at-point :around #'cape-wrap-purify))
  :defer t)
#+end_src
```

**Comment by Jdogzz** [comment]
The linked bug report indicates this may have been fixed as of emacs 29.1: https://debbugs.gnu.org/cgi/bugreport.cgi?bug=40651

**Comment by hlissner** [comment]
I would like to avoid the extra dependency, so I've tried eyeballing `cape-wrap-silent` and `cape-wrap-purify` into two self-contained advice. Try dropping it into `$DOOMDIR/config.el` and let me know if they resolve the issue:

```elisp
(after! eshell
  (defadvice! +eshell--silent-a (capf)
    :around #'pcomplete-completions-at-point
    (cl-macrolet ((silent (&rest body) `(quiet! (ignore-errors ,@body)))
                  (wrapped-table (wrap body)
                    `(lambda (str pred action)
                       (,@body
                        (let ((result (complete-with-action action table str pred)))
                          (when
                              (and (eq action 'completion--unquote)
                                   (functionp (cadr result)))
                            (cl-callf ,wrap (cadr result)))
                          result)))))
      (cl-labels ((silent-table (table) (wrapped-table silent-table (silent))))
        (pcase (silent (funcall capf))
          (`(,beg ,end ,table . ,plist)
           `(,beg ,end ,(silent-table table) ,@plist))))))

  (defadvice! +eshell--purify-a (capf)
    :around #'pcomplete-completions-at-point
    (catch 'illegal-completion-in-region
      (condition-case nil
          (let ((buffer-read-only t)
                (inhibit-read-only nil)
                (completion-in-region-function
                 (lambda (beg end coll pred)
                   (throw 'illegal-completion-in-region
                          (list beg end coll :predicate pred)))))
            (funcall capf))
        (buffer-read-only nil)))))
```

**Comment by hlissner** [comment]
I've merged this fix in 8340141, which appears to resolve the issue. Let me know if that isn't the case and I'll reopen this. Thanks for bringing it to my attention!
